Automattic/wp-calypso
# Checkout: decouple useCreatePaymentCompleteCallback and useGetThankYouUrl from CompositeCheckout

#### Changes proposed in this Pull Request

In #48068 we moved the creation of checkout's `onPaymentComplete` function into its own hook, but it still has a few dependencies on the environment created by `CompositeCheckout`. Particularly, it relies on the checkout analytics handler to record the purchase (via `recordEvent`) and it uses the "wpcom" `@wordpress/data` store. The related hook `useGetThankYouUrl` also uses the "wpcom" store.

In this PR we simplify the use of `useCreatePaymentCompleteCallback` by moving the analytics calls into the hook directly, meaning that the hook no longer requires as much data and it is even more decoupled from composite checkout.

Secondly we move the responsibility of fetching the transaction result out of both hooks, which allows us to decouple them from the composite checkout `@wordpress/data` store. Previously when the transaction was completed for every payment method, a side effect was triggered to save the transaction data in the "wpcom" store so it could be pulled out later. It turns out this is unnecessary because the transaction data is already being saved in the `useTransactionStatus` hook in composite checkout, so this PR removes all those side effects and changes the `onPaymentComplete` handler so it also receives the transaction result.

These changes will make it possible to use these hooks in the one-click upsell modal as that is converted from old checkout (see #48132).

#### Testing instructions

- Make sure you are recording network activity (or debug logs for analytics) so you can see Tracks events firing and make sure that those logs will persist after a redirect.
- Visit checkout with a product in your cart (preferably a Personal or Premium plan to get the upsells).
- Complete checkout with a non-redirect payment method (eg: credit card or saved card).
- Verify that there are no errors and that you are redirected to a thank-you upsell page.
- Verify that the receipt ID in the thank-you page URL is correct. This is very important as it proves that the transaction result was passed in correctly.
- Verify that the `calypso_checkout_payment_success` Tracks event is created and nothing looks out-of-place (like null values or something).
- Repeat the above steps for a redirect payment method (eg: PayPal). (Note that you will not see the `calypso_checkout_payment_success` event; this is a known issue.)
